### PR TITLE
Add printing support

### PR DESCRIPTION
**Pull request by henke37**
This aims to implement printing.

Initial support is limited to HE Scumm titles and the Windows backend.

There are things left to iron out before a merge:
- [ ] Code style issues
- [x] Use `ManagedSurface`
- [x] Colors are lost, everything is black and white
- [x] Output doesn't fill the page
- [ ] Support more Wiz compression modes.
- [ ] `testbed` tests
- [x] Build config feature flag

Known issues that are not my fault:
* Fonts are not implemented in HE Scumm
* Crash when loading a saved document in HE Scumm
* Multi page printing creates a print job for each page. That's how the original interpreter works.

### CODE REVIEW

File: backends/printing/printman.h [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+#include "common/scummsys.h"
+#include "common/str.h"
+
+class PrintingManager {
+public:
+	virtual ~PrintingManager() = 0;
+
+	virtual void printImage(Common::String jobName, byte *pixels, byte *palette, uint32 width, uint32 height) = 0;
+	void printImage(byte * pixels, byte * palette, uint32 width, uint32 height) {
```
Comment (sev-) [inline backends/printing/printman.h:31]: This API is extremely limited. Basically, here you're printing only 8 bit images which is way too limited.

My suggestion is to pass here a `Graphics::ManagedSurface()` which could be of any format and even a palette attached to it.

Comment (henke37) [reply backends/printing/printman.h:31]: I thought I looked at that. But I will take a second look. I'm not experienced with the graphics subsystem, so I must've overlooked it.

Comment (sev-) [reply backends/printing/printman.h:31]: `Graphics::ManagedSurface()` has `setPalette()` method.

In this particular case, with the WIZ images, you need to create a CLUT8 surface, copy over the pixels and set the palette. Should be like 2 loops and a single call

Comment (henke37) [reply backends/printing/printman.h:31]: Yeah, I'm seeing that now.

Btw, you happen to know the comp == 1 format for wiz images? I can't make heads or tails out of it. The Let's Explore series is printing in that format.

**Comment by sev-** [comment]
For features like this it is very important to create a test in the testbed/ engine right away. So, the backend developers have possibility to test it.

Also, please add a backend feature, see `enum Feature` in `common/system.h` and check for the presence of this feature with `g_system->q()` and show a message with `GUI::MessageDialog` if printing is not supported by the backend.

File: backends/module.mk [hunk 2]
```diff
@@ -25,7 +25,8 @@ MODULE_OBJS := \
 	midi/timidity.o \
 	saves/savefile.o \
 	saves/default/default-saves.o \
-	timer/default/default-timer.o
+	timer/default/default-timer.o \
+	printing/printman.o
```
Comment (sev-) [inline backends/module.mk:29]: This has to be sorted alphabetically

File: backends/module.mk [hunk 3]
```diff
@@ -242,7 +243,8 @@ MODULE_OBJS += \
 	plugins/win32/win32-provider.o \
 	saves/windows/windows-saves.o \
 	updates/win32/win32-updates.o \
-	taskbar/win32/win32-taskbar.o
+	taskbar/win32/win32-taskbar.o \
+	printing/win32/win32-printman.o
```
Comment (sev-) [inline backends/module.mk:247]: This has to be sorted alphabetically

**Comment by henke37** [comment]
Right, update the `testbed`. I'll make sure to whip a test. Good thing that there's a handy logo to print.

And with the dialog in place, would it be appropriate to remove the ini overrides?

File: backends/printing/printman.h (hunk 1 above)
Comment (sev-) [reply backends/printing/printman.h:31]: Did you see line 1621 in `wiz_he.cpp`?  https://github.com/scummvm/scummvm/blob/master/engines/scumm/he/wiz_he.cpp#L1621

File: backends/printing/win32/win32-printman.h [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#ifndef BACKENDS_PRINTMAN_WIN32_H
```
Comment (sev-) [inline backends/printing/win32/win32-printman.h:22]: The guard define should match the file name

**Comment by sev-** [comment]
> And with the dialog in place, would it be appropriate to remove the ini overrides?

Which ini overrides? Please, elaborate

**Review by lephilousophe — COMMENTED** [review]
I have multiple comments on this PR but I forgot to send them yesterday.

In addition I wonder if the API defined is good enough.
It only takes care of CLUT8 format and no other.

Moreover, a basic implementation outputting images to disk would be a minimum before using the Win32 printing API.

**Comment by henke37** [comment]
> > And with the dialog in place, would it be appropriate to remove the ini overrides?
> 
> Which ini overrides? Please, elaborate

`ScummEngine_v72he::o72_readINI` and `ScummEngine_v70he::o70_readINI` have special handling for `NoPrinting`.

File: backends/printing/printman.h (hunk 1 above)
Comment (henke37) [reply backends/printing/printman.h:31]: I did see that one yes. Sadly didn't make things clearer. I will take a third look at it later.

File: backends/printing/printman.cpp [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "printman.h"
+
+PrintingManager::~PrintingManager() {}
```
[the remark this answers is not in the document]
Comment (henke37) [reply backends/printing/printman.cpp:24]: I hope to make the base class more substantial as the api stabilizes.

File: backends/printing/printman.h [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+
+#include "common/scummsys.h"
+#include "common/str.h"
+
+class PrintingManager {
+public:
+	virtual ~PrintingManager() = 0;
+
+	virtual void printImage(Common::String jobName, byte *pixels, byte *palette, uint32 width, uint32 height) = 0;
```
[the remark this answers is not in the document]
Comment (henke37) [reply backends/printing/printman.h:30]: Fixed locally, and soon to be removed with the ManagedSurface refactoring.

File: backends/printing/win32/win32-printman.cpp [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+void Win32PrintingManager::printImage(Common::String jobName, byte *pixels, byte *paletteData, uint32 width, uint32 height) {
+	HDC hdcPrint = createDefaultPrinterContext();
+
+	HDC hdcImg = CreateCompatibleDC(hdcPrint);
+	
+	HPALETTE pal = buildPalette(paletteData);
+	HBITMAP bitmap = buildBitmap(hdcImg, pixels, paletteData, width, height);
+	if (!bitmap)
+		goto delDC;
```
[the remark this answers is not in the document]
Comment (henke37) [reply backends/printing/win32/win32-printman.cpp:57]: I'll write some handle wrapper classes then.

File: backends/printing/win32/win32-printman.cpp [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+	DeleteDC(hdcImg);
+	DeleteDC(hdcPrint);
+}
+
+HDC Win32PrintingManager::createDefaultPrinterContext() {
+	wchar_t szPrinter[MAX_PATH];
+	BOOL success;
+	DWORD cchPrinter(ARRAYSIZE(szPrinter));
+
+	success=GetDefaultPrinterW(szPrinter, &cchPrinter);
```
[the remark this answers is not in the document]
Comment (henke37) [reply backends/printing/win32/win32-printman.cpp]: Done.

File: backends/printing/win32/win32-printman.cpp [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+	ClosePrinter(handle);
+
+	HDC printerDC = CreateDCW(L"WINSPOOL", devName, NULL, devmode);
+	return printerDC;
+}
+
+HPALETTE Win32PrintingManager::buildPalette(byte *paletteData) {
+	LOGPALETTE *lpal = (LOGPALETTE*)malloc(sizeof(LOGPALETTE) + (256-1)*sizeof(PALETTEENTRY));
+
+	if (!lpal)
```
[the remark this answers is not in the document]
Comment (henke37) [reply backends/printing/win32/win32-printman.cpp:147]: Not sure if `new` is appropriate when you need to concatenate structs like the WinApi likes.

File: backends/printing/printman.h [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+
+#include "common/scummsys.h"
+#include "common/str.h"
+
+class PrintingManager {
```
[the remark this answers is not in the document]
Comment (henke37) [reply backends/printing/printman.h:29]: Do you still think that the `Common` namespace is correct now that I've introduced the `PrintJob` class?

**Comment by sev-** [comment]
> > > And with the dialog in place, would it be appropriate to remove the ini overrides?
> > 
> > 
> > Which ini overrides? Please, elaborate
> 
> `ScummEngine_v72he::o72_readINI` and `ScummEngine_v70he::o70_readINI` have special handling for `NoPrinting`.

Yes, sure, just guard those with USE_PRINTING

File: backends/printing/win32/win32-printman.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (sev-) [reply backends/printing/win32/win32-printman.cpp:90]: or you could write
```C++
if (!bitmap) {
   DeleteDC(hdcImg);
   return;
}
```

We do that a lot when we construct objects or call malloc()

File: engines/scumm/he/wiz_he.cpp [hunk 11]
```diff
@@ -1491,7 +1492,29 @@ uint8 *Wiz::drawWizImage(int resNum, int state, int maskNum, int maskState, int
 	}
 
 	if (flags & kWIFPrint) {
-		error("WizImage printing is unimplemented");
+		switch (comp) {
+		case 0: {
+			Graphics::ManagedSurface surf;
+			surf.w = width;
+			surf.h = height;
+			surf.format=Graphics::PixelFormat::createFormatCLUT8();
```
Comment (sev-) [inline engines/scumm/he/wiz_he.cpp:1500]: Formatting.

File: backends/printing/win32/win32-printman.cpp [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+	BITMAPINFO *bitmapInfo = (BITMAPINFO *)malloc(sizeof(BITMAPINFO) + sizeof(RGBQUAD) * (colorCount - 1));
+
+	if (!bitmapInfo)
+		return NULL;
+
+	bitmapInfo->bmiHeader.biSize = sizeof(BITMAPINFOHEADER);
+	bitmapInfo->bmiHeader.biWidth = surf.w;
+	bitmapInfo->bmiHeader.biHeight = -((LONG)surf.h); // Blame the OS2 team for bitmaps being upside down
+	bitmapInfo->bmiHeader.biPlanes = 1;
+	bitmapInfo->bmiHeader.biBitCount = (surf.format.isCLUT8()?8:surf.format.bpp());
```
Comment (sev-) [inline backends/printing/win32/win32-printman.cpp:155]: Formatting

File: engines/scumm/he/wiz_he.cpp [hunk 13]
```diff
@@ -1491,7 +1492,29 @@ uint8 *Wiz::drawWizImage(int resNum, int state, int maskNum, int maskState, int
...
+			assert(pal);
+			surf.setPalette(pal, 0, 256);
+
+			PrintingManager *pm = _vm->_system->getPrintingManager();
+			pm->printImage(surf);
+
+			return 0;
+		}
+		default:
+			error("Unsupported compression mode %d", comp);
```
Comment (sev-) [inline engines/scumm/he/wiz_he.cpp:1516]: Let's talk about the  compmode 1 here.

Probably, you do not need to understand it, just call those unpackers here and they will return you a normal bitmap. E.g. just copy over the calls and checks for masks.

File: engines/testbed/printing.cpp [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+#include "base/version.h"
+
+#include "common/events.h"
+#include "common/keyboard.h"
```
Comment (sev-) [inline engines/testbed/printing.cpp:24]: I think this file has unneeded includes.

File: engines/testbed/printing.cpp [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+
+
+TestExitStatus PrintingTests::printTestPage() {
+	if (!ConfParams.isSessionInteractive()) {
+		return kTestSkipped;
+	}
+	if (!g_gui.theme()->supportsImages())
+		return kTestSkipped;
+	const Graphics::ManagedSurface *logo = g_gui.theme()->getImageSurface("logo.bmp");
+	if (!logo)
```
Comment (sev-) [inline engines/testbed/printing.cpp:71]: Please issue a `warning()` here

File: engines/testbed/printing.cpp [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+		return kTestFailed;
+	
+	PrintingManager *pm = g_system->getPrintingManager();
+	if (!pm)
+		return kTestFailed;
```
Comment (sev-) [inline engines/testbed/printing.cpp:76]: And here. So the developers know why the test has failed without studying the whole test codebase.

File: backends/printing/printman.h [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+		printImage("ScummVM", surf);
+	}
+};
+
+class PrintJob {
+public:
+	friend class PrintingManager;
+	virtual ~PrintJob();
+
+	virtual void drawBitmap(const Graphics::ManagedSurface &surf, int x, int y) = 0;
```
Comment (sev-) [inline backends/printing/printman.h:50]: Maybe `printImage()` is still more logical name here?

File: engines/testbed/printing.cpp [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+TestExitStatus PrintingTests::printGPL() {
+	if (!ConfParams.isSessionInteractive()) {
+		return kTestSkipped;
+	}
+
+	PrintingManager *pm = g_system->getPrintingManager();
+	if (!pm)
+		return kTestFailed;
+
+	PrintJob *job = pm->createJob("The GPL");
```
Comment (sev-) [inline engines/testbed/printing.cpp:101]: Try `file.open("COPYING")`  And if it fails, just print some static text like "The quick brown fox jumps over the lazy dog"

File: engines/testbed/printing.h [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#ifndef TESTBED_PRINTING_H
+#define TESTBED_PRINTING_H
+
+#include "testbed/testsuite.h"
+
+// This file can be used as template for header files of other newer testsuites.
```
Comment (sev-) [inline engines/testbed/printing.h:27]: A leftover comment

**Review by sev- — CHANGES_REQUESTED** [review]
Did another round of reviews

File: backends/printing/printman.h [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+ *
+ */
+
+
+#include "common/scummsys.h"
+#include "common/str.h"
+
+class PrintingManager {
+public:
+	virtual ~PrintingManager() = 0;
```
[the remark this answers is not in the document]
Comment (lephilousophe) [reply backends/printing/printman.h:34]: This is still not a pure virtual: you declared a destructor.

File: backends/printing/printman.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (lephilousophe) [reply backends/printing/printman.h:32]: `Common` or `Graphics`, yes.
Currently it has no namespace.

File: backends/printing/win32/win32-printman.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (lephilousophe) [reply backends/printing/win32/win32-printman.cpp:90]: Yes, exactly what @sev- said.
Keep it simple.
For every allocation, a check and if a failure, an unwrap of all the allocations done above.
No need of RAII for such lifetimes.

File: backends/printing/printman.h [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+#include "common/str.h"
+#include "graphics/managed_surface.h"
+
+class PrintJob;
+
+class PrintingManager {
+public:
+	virtual ~PrintingManager() = 0;
+
+	virtual PrintJob *createJob(Common::String jobName) = 0;
```
Comment (lephilousophe) [inline backends/printing/printman.h:36]: Use a const reference here.

File: backends/printing/printman.h [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+
+	void printImage(Common::String jobName, const Graphics::ManagedSurface &surf);
```
Comment (lephilousophe) [inline backends/printing/printman.h:38]: Same here : const reference.

**Comment by lephilousophe** [comment]
I don't understand why you now have `PrintManager` and `PrintJob` with both a subclass specific for each platform.
`PrintManager` is a simple factory for `PrintJob` objects but there is already a factory: `OSystem`.
It's simpler to have a `createJob()` function in `OSystem` which will return a new `PrintJob`.
This function would return `nullptr` when not supported.
It also prevents having you a `PrintManager` instance lying around when there is almost never a use for it.

About the job name, I would go to have a setter for it instead of setting it in the constructor.
This avoids to have 2 functions.

File: engines/testbed/printing.cpp (hunk 14 above)
Comment (henke37) [reply engines/testbed/printing.cpp:24]: `version.h` is supposed to be used for the basic test page. But the others are indeed unneeded.

**Comment by henke37** [comment]
The distinction between PrintingManager and PrintJob is that _well designed_ code will check with the user for how to print before starting a job. Stuff like checking paper type and adjusting the on screen layout based on the result. And maybe even checking the printer capabilities, such as supporting color output.

**Comment by sev-** [comment]
> The distinction between PrintingManager and PrintJob is that _well designed_ code will check with the user for how to print before starting a job. Stuff like checking paper type and adjusting the on screen layout based on the result. And maybe even checking the printer capabilities, such as supporting color output.

No way. Please, do not over-engineer it for some features which will never be implemented.

File: backends/printing/win32/win32-printman.cpp [hunk 23]
```diff
@@ -109,6 +119,48 @@ void Win32PrintJob::setTextColor(int r, int g, int b) {
 	SetTextColor(hdcPrint, RGB(r, g, b));
 }
 
+Common::Rect Win32PrintJob::getTextBounds(const Common::String &text) const {
+	SIZE winSize;
+	GetTextExtentPointA(hdcPrint, text.c_str(), text.size(), &winSize);
+	return Common::Rect(winSize.cx, winSize.cy);
+}
+
+Common::Rational Win32PrintJob::getPixelAspectRatio() const {
```
Comment (sev-) [inline backends/printing/win32/win32-printman.cpp:128]: Do you have any use-case for all these parameters? E.g. how do you think it could used by engines and by which specifically?

Comment (henke37) [reply backends/printing/win32/win32-printman.cpp:128]: I don't have use cases for all of them. But the new `testbed` code shows some of them in use.

**Comment by lephilousophe** [comment]
I persist in saying that before adding more features to Win32, an implementation outputting image files (or PDF if it's easy and not too heavy) should be done.
That's the most useful use case nowadays as printers are becoming rare (and it's more eco-friendly).

In addition a way to specify printer name should be done instead of blindly selecting the default one.
All of this is a must have before adding features yet to be used by engines.

**Comment by henke37** [comment]
Most oses come with a print to pdf driver. It's what I'm using for my testing.

As for specifying the printer name, I agree that it should be possible. But IIRC, that step is already mandatory on both ios and android. You can't avoid it on those platforms.

While at the same time, at least HE games don't have that option in the original interpreter. This would be an intentional deviation from the original.

**Comment by sev-** [comment]
> While at the same time, at least HE games don't have that option in the original interpreter. This would be an intentional deviation from the original.

Excuse me, but how? If you allow selecting printer in the config, how will it deviate? And do I read you correctly that you are designing this feature around a single engine?

**Comment by henke37** [comment]
At least on Windows it is an entirely optional step to show the print settings dialog box. That's what I was talking about.

As for selecting the printer in the config, it is IIRC, impossible to do on android and ios. You choose the printer each time you print. Perhaps a windows specific solution can be made. I'm not ruling that out. But when the os already has the concept of a default printer, is there much to gain from a separate setting in ScummVM? A duplicate risks confusing users and won't keep up with printers being added and removed to the system.

And no, I'm not designing around a single engine. I'm just limiting the initial implementation to what I can test.

**Comment by henke37** [comment]
This last commit breaks the build due to a missing import library. I will address that at the same time as I take on the build feature stuff.

File: backends/printing/win32/win32-printman.cpp (hunk 23 above)
Comment (sev-) [reply backends/printing/win32/win32-printman.cpp:195]: Then please remove them as they are not used by real engines.

File: backends/printing/printman.h [hunk 24]
```diff
@@ -0,0 +1,107 @@
...
+#include "common/file.h"
+#include "graphics/managed_surface.h"
+
+class PrintJob;
+
+class PrintingManager {
+public:
+	virtual ~PrintingManager();
+
+	virtual PrintJob *createJob(const Common::String &jobName) = 0;
```
Comment (sev-) [inline backends/printing/printman.h:40]: Could you please explain the reasoning behind this PrintJob class?

Comment (henke37) [reply backends/printing/printman.h:40]: I believe I already did? Let me try again then.

The PrintingManager class is the central starting point for the feature of printing. It is where things such as printer enumeration should go.

The PrintJob class represents an active print job that is being spooled. The user has decided to print and with what settings to do so.

File: backends/printing/win32/win32-printman.cpp (hunk 23 above)
Comment (henke37) [reply backends/printing/win32/win32-printman.cpp:195]: The way I see it, this is a complicated matter.

A feature added now is a burden on all backends that want to implement this. But a feature *not* added now that turns out to be needed later is an even bigger challenge. Editing the interface by adding member functions will cause existing implementations to break. Breaking existing stuff is very uncool.

As such I've been using some guidelines for what does and does not make it:
* Is the feature portable? If win, ios and android apis all have it its much more likely to make it than if it is just one of them.
* Is the feature complex to implement? A simple getter that directly queries the platform api is much simpler than something that would require bringing in hundreds, if not thousands, of lines of code.
* Is there a reasonable default if it can't be implemented? Hardcoding a return value that makes sense is much better than having to call `error`.

File: backends/printing/printman.h (hunk 24 above)
Comment (lephilousophe) [reply backends/printing/printman.h:40]: But there is nothing here that can't be done in PrintJob.
This would make the implementation way simpler.

**Comment by sev-** [comment]
@henke37 please, implement these requests of @lephilousophe and mine:

- The API has to be very simple: print image, that's it.
- No PrintJob class
- A way to save the requested image into a file

**Comment by henke37** [comment]
I have to disagree. This api needs to account for future users. Just printing a bitmap is not enough. I wouldn't mind such an initial limit, if it wasn't for the fact that it can't be extended later. Backend implementers need a stable api that does not get surprise additions later.

In fact, I've been pondering how to add some vector drawing features. Printing is done at a high resolution and vector graphics excel at that. It's already obvious in my testing that bitmaps look bad upscaled. I'm thinking of the `VectorRenderer` class here, it seems about right in scope.

Saving the result to a file is a good idea. But I maintain that platforms already have it built in. And they aren't limited to a bitmap. They've taken the time to handle all output, such as vector graphics and text. Writing code that can handle that is more trouble than I'm ready to bit off.

**Comment by sev-** [comment]
> I have to disagree.

No problem. I am closing this pull request then.

**Comment by mikrosk** [comment]
> Saving the result to a file is a good idea. But I maintain that platforms already have it built in.

I could name you half a dozen which do not. :)
